src/adb.js

```diff
@@ -1 +1,393 @@
-// https://github.com/appcelerator/titanium_mobile/blob/master/node_modules/titanium-sdk/lib/adb.js
+import { EventEmitter } from 'events';
+import fs from 'fs';
+import mkdirp from 'mkdirp';
+import path from 'path';
+import { spawn } from 'child_process';
+
+import { Connection } from './connection';
+import { EmulatorManager } from './emulator';
+import * as sdk from './sdk';
+import * as util from './util';
+
+/**
+ * Creates an ADB object.
+ *
+ * @class
+ * @extends EventEmitter
+ * @classdesc Provides methods to interact with the Android Debug Bridge (ADB).
+ * @constructor
+ */
+export class ADB extends EventEmitter {
+	constructor(opts = {}) {
+		super();
+		this.opts = opts;
+	}
+
+	/**
+	 * Returns the version of the ADB server.
+	 *
+	 * @returns {Promise}
+	 * @access public
+	 */
+	version() {
+		return new Connection(this)
+			.exec('host:version')
+			.then(data => '1.0.' + parseInt(data, 16));
+	}
+
+	/**
+	 * Runs the specified command on the Android emulator/device.
+	 *
+	 * @param {String} deviceId - The id of the device or emulator.
+	 * @param {String} cmd - The command to run.
+	 * @returns {Promise}
+	 * @access public
+	 */
+	shell(deviceId, cmd) {
+		if (typeof deviceId !== 'string' || !deviceId) {
+			return Promise.reject(new TypeError('Expected device ID to be a string.'));
+		}
+
+		if (typeof cmd !== 'string' || !cmd) {
+			return Promise.reject(new TypeError('Expected command to be a string.'));
+		}
+
+		const conn = new Connection(this);
+		return conn
+			.exec('host:transport:' + deviceId)
+			.then(data => conn.exec(`shell: ${cmd.replace(/^shell\:/, '')}`, { bufferUntilClose: true, noLength: true }));
+	}
+
+	/**
+	 * Attempts to find the adb executable, then start the adb server.
+	 *
+	 * @returns {Promise}
+	 * @access public
+	 */
+	startServer() {
+		return sdk
+			.detect(this.opts)
+			.then(result => util.run(result.sdk.executables.adb, ['start-server']));
+	}
+
+	/**
+	 * Attempts to find the adb executable, then stop the adb server.
+	 *
+	 * @returns {Promise}
+	 * @access public
+	 */
+	stopServer() {
+		return sdk
+			.detect(this.opts)
+			.then(result => util.run(result.sdk.executables.adb, ['kill-server']));
+	}
+
+	/**
+	 * Forwards the specified device/emulator's socket connections to the destination.
+	 *
+	 * @param {String} deviceId - The id of the device or emulator.
+	 * @param {String} src - The source port in the format "tcp:<port>".
+	 * @param {String} dest - The destination port in the format "tcp:<port>" or "jdwp:<pid>".
+	 * @returns {Promise}
+	 * @access public
+	 */
+	forward(deviceId, src, dest) {
+		if (typeof deviceId !== 'string' || !deviceId) {
+			return Promise.reject(new TypeError('Expected device ID to be a string.'));
+		}
+
+		src = util.expandPath(src);
```

Comment:
What happens if `src` is not a string? You should check that `src` is a string and that it isn't empty.